### PR TITLE
feat(cdc) Pre filter cdc messages based on table header [take 2]

### PR DESCRIPTION
**Pull request by fpacifici**
The goal is to drop cdc messages before the json payload gets parsed if the Kafka headers contain a table the consumer is not interested into.

The current  Consumer is not doing any dataset specific operation before json parsing and I would not want to move parsing into the MessageProcessor (since this would make the processor Kafka specific while now they are fairly agnostic to the stream type).
I added, instead, a pre-filter component (provided by the StreamLoader (which is provided by the Table Writer) together with the MessageProcessor). This is Kafka specific, if the stream is a Kafka stream, and it makes a step in making the MessageProcessor fully agnostic to the stream type.

### CODE REVIEW

File: snuba/datasets/cdc/cdcprocessors.py [hunk 1]
```diff
@@ -116,11 +112,6 @@ def _process_delete(
     def process_message(self, value, metadata) -> Optional[ProcessedMessage]:
         assert isinstance(value, dict)
 
-        partition = metadata.partition
-        assert (
-            partition == KAFKA_ONLY_PARTITION
-        ), "CDC can only work with single partition topics for consistency"
-
```
Comment (fpacifici) [inline snuba/datasets/cdc/cdcprocessors.py:123]: This takes a step in making the MessageProcessor fully Kafka agnostic. Now we still have the offset lingering around in the KafkaMetadata, but that's the last thing to remove.

File: snuba/consumer.py [hunk 2]
```diff
@@ -48,9 +48,12 @@ def __init__(
...
-        # TODO: consider moving this inside the processor so we can do a quick
-        # processing of messages we want to filter out without fully parsing the
-        # json.
+        pre_filter = (
+            enforce_table_writer(self.__dataset).get_stream_loader().get_pre_filter()
+        )
+        if pre_filter and pre_filter.should_drop(message):
+            return None
+
         if self.__rapidjson_deserialize:
```
Comment (fpacifici) [inline snuba/consumer.py:57]: I did not make the parsing phase storage specific as well because, as of now, none needs that, though we could easily go that way and actually make the storage provide all the components used by this class to process a message.

File: snuba/datasets/table_storage.py [hunk 3]
```diff
@@ -31,6 +33,7 @@ def __init__(
         self,
         processor: MessageProcessor,
         default_topic: str,
+        pre_filter: Optional[StreamMessageFilter[KafkaPayload]] = None,
```
Comment (fpacifici) [inline snuba/datasets/table_storage.py:36]: This class itself is still Kafka specific. We could build an abstract one on top of it. IT would not be used today.

File: snuba/consumer.py [hunk 4]
```diff
@@ -48,9 +48,12 @@ def __init__(
     def process_message(
         self, message: Message[KafkaPayload]
     ) -> Optional[ProcessedMessage]:
-        # TODO: consider moving this inside the processor so we can do a quick
-        # processing of messages we want to filter out without fully parsing the
-        # json.
+        pre_filter = (
+            enforce_table_writer(self.__dataset).get_stream_loader().get_pre_filter()
+        )
```
Comment (tkaemming) [inline snuba/consumer.py:53]: Should this be moved to a private instance variable instead of performing this lookup chain every time?

File: snuba/consumer.py (hunk 2 above)
Comment (tkaemming) [reply snuba/consumer.py:57]: This comment was made prior to the discussion of entities — my understanding is approach wouldn't make as much sense under those circumstances, since the consumer will be transforming messages to entities, which are then transformed to storage rows/operations based on the writable storage implementation?

File: snuba/datasets/cdc/message_filters.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        self.__postgres_table = postgres_table
+
+    def should_drop(self, message: Message[KafkaPayload]) -> bool:
+        assert (
+            message.partition.index == KAFKA_ONLY_PARTITION
+        ), "CDC can only work with single partition topics for consistency"
+
+        table_header = next(
+            (header for header in message.payload.headers if header[0] == "table"), None
+        )
```
Comment (tkaemming) [inline snuba/datasets/cdc/message_filters.py:31]: Since the header key isn't used anywhere below, it seems like it'd make more sense to do this:

```suggestion
        table_name = next(
            (value for key, value in message.payload.headers if key == "table"), None
        )
```

File: snuba/datasets/cdc/message_filters.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+        if table_header:
+            table_name = table_header[1].decode("utf-8")
+            if table_name != self.__postgres_table:
+                metrics.increment("CDC Message Dropped", tags={"table": table_name})
```
Comment (tkaemming) [inline snuba/datasets/cdc/message_filters.py:36]: Does datadog actually allow string with spaces? I've never looked, but it seems to go against our typical usage pattern.

File: tests/datasets/test_groupassignee.py [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import pytz
-import simplejson as json
+import rapidjson
```
Comment (tkaemming) [inline tests/datasets/test_groupassignee.py:2]: This seems unrelated? I know there was talk about enforcing that as a default earlier, but that doesn't seem to have made it into this change.

File: tests/test_groupedmessage.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import pytz
-import simplejson as json
+import rapidjson
```
Comment (tkaemming) [inline tests/test_groupedmessage.py:2]: Same comment about the JSON change.

**Review by tkaemming — APPROVED** [review]
Some nits inline but nothing blocking. The metrics one should probably be addressed for consistency's sake if nothing else, though.

File: snuba/consumer.py (hunk 2 above)
Comment (fpacifici) [reply snuba/consumer.py:57]: I think the consumer will change even more with entities. Yes your pipeline seems the one we discussed. I think a concept of pre-filtering and then parsing can still be valid there even though it would look different.

File: snuba/datasets/cdc/message_filters.py (hunk 6 above)
Comment (fpacifici) [reply snuba/datasets/cdc/message_filters.py:36]: good point. statsd did, datadog may not.

File: tests/datasets/test_groupassignee.py (hunk 7 above)
Comment (fpacifici) [reply tests/datasets/test_groupassignee.py:2]: it is indeed unrelated.
It came from the first attempt: https://github.com/getsentry/snuba/pull/857 where it was needed
when I copied over the change to this file I forgot to remove that change

File: tests/test_groupedmessage.py (hunk 8 above)
Comment (fpacifici) [reply tests/test_groupedmessage.py:2]: same answer
